File: neetcode150/medium/33-Search-in-Rotated-Sorted-Array.py
# Solution 1
class Solution:
  def search(self, nums, target):
    return self.binary(nums, 0, len(nums) - 1, target)
    
  def binary(self, nums, begin, end, key):
    if begin > end:
      return -1
        
    mid = (end + begin) // 2
        
    if nums[mid] == key:
      return mid
        
    elif nums[begin] < nums[mid] and key <= nums[mid] and key >= nums[begin]:
      return self.binary(nums, begin, mid - 1, key)
    elif nums[mid] < nums[end] and key >= nums[mid] and key <= nums[end]:
      return self.binary(nums, mid + 1, end, key)
    elif nums[end] < nums[mid]:
      return self.binary(nums, mid + 1, end, key)
    elif nums[begin] > nums[mid]:
      return self.binary(nums, begin, mid - 1, key)
    return -1
           

File: neetcode150/medium/33-Search-in-Rotated-Sorted-Array.cpp
#include <vector>
using namespace std;

// Solution 1
class Solution {
public:
  int search(vector<int>& nums, int target) {
    int begin = 0;
    int end = nums.size() - 1;
    while(begin <= end){
      int mid = (begin + end) / 2;
      // Check if the middle value is equal to target
      if(nums[mid] == target){
        return mid;
      } if(nums[begin] > nums[mid] && target >= nums[mid] && target >= nums[begin]){
        end = mid - 1;
      } else if(nums[end] < nums[mid] && target <= nums[mid] && target <= nums[end]){
        begin = mid + 1;
      // Target greater than mid search the upper half of the list
      } else if(target > nums[mid]){
        begin = mid + 1; 
      // Target less than mid search the lower half of the list
      } else if(target < nums[mid]){
        end = mid - 1; 
      }
    }
    return -1;
  }
};